Add NRGkick integration and tests

andijakl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR brings the [NRGkick](https://nrgkick.com/en/) mobile EV charger (wallbox) integration from HACS ( [GitHub repo](https://github.com/andijakl/nrgkick-homeassistant) ) to HA Core. It supports reading the state using the official local REST API (accessed via PyPi library).

To comply with best practices, I've trimmed down my existing community integration to just the sensor platform and will add charging control back in future updates to the HA core integration.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/42806
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

andijakl committed:
Add nrgkick community integration as ha core integration

andijakl committed:
Improve error messages

andijakl committed:
Improve translations and strict typing

andijakl committed:
Add dynamic thresholds for setting charging current

andijakl committed:
Limit nrgkick to sensor platform

andijakl committed:
Add quality scale evidence links

andijakl committed:
Remove configurable polling interval

andijakl committed:
Remove quality scale progress for silver and gold tier, only report bronze status for first pull request

andijakl committed:
Add discovered device IP to zeroconf confirmation step in translation

joostlek commented on homeassistant/components/nrgkick/api.py:
Have you considered using a decorator on the methods where you call the api? That might be a less invasive option of wrapping the calls

joostlek commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,290 @@
+    }
+
+
+# pylint: disable=abstract-method  # is_matching is not required for HA config flows
```
Can we get rid of all the pylint disables?

joostlek commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,290 @@
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Optional(CONF_USERNAME): str,
+        vol.Optional(CONF_PASSWORD): str,
```
why are they optional?

joostlek replied:
Can we detect when we need it? This sounds like it's off by default, so in theory we could just
1. user flow, ask for the host, try to connect, if it doesn't work, ask for the auth credentials
2. zeroconf flow, try to connect underwater, if it works, we just ask the user if they want to add it
3. if that connection failed because auth, we can ask the user for credentials whenever they press the button.

This way if you don't have credentials setup, you're never asked to enter them.

andijakl replied:
Good idea - added the config flow steps for user & zeroconf flow to first check without auth and only ask if necessary. Will perform extensive tests with the real device before resolving this.

joostlek commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,290 @@
+
+    return {
+        "title": device_name,
+        "serial": info.get("general", {}).get("serial_number", "Unknown"),
```
when would there be no serial number? As this would be the unique id of the entry, I don't like the idea of just setting it to Unknown. When could this happen?

joostlek commented on homeassistant/components/nrgkick/coordinator.py:
```diff
@@ -0,0 +1,202 @@
+            "info": info,
+            "control": control,
+            "values": values,
+        }
```
Consider using a dataclass to ensure typesafety for these parts

andijakl replied:
Yes, makes sense - thanks for the suggestion, now implemented!

joostlek commented on homeassistant/components/nrgkick/diagnostics.py:
Let's keep for a followup

joostlek commented on homeassistant/components/nrgkick/entity.py:
```diff
@@ -0,0 +1,57 @@
+        data = self.coordinator.data
+        info_data: dict[str, Any] = data.get("info", {}) if data else {}
+        device_info: dict[str, Any] = info_data.get("general", {})
+        serial: str = device_info.get("serial_number", "unknown")
```
you can add the serial number to the device info, and again, we shouldn't fallback to unknown

joostlek commented on homeassistant/components/nrgkick/entity.py:
```diff
@@ -0,0 +1,57 @@
+
+        device_name: str | None = device_info.get("device_name")
+        if not device_name:
+            device_name = "NRGkick"
```
this is already used for the entry title. If you set the device name to None, it will use the entry.title for the device name

andijakl replied:
True, that was redundant. Now fixed!

joostlek commented on homeassistant/components/nrgkick/quality_scale.yaml:
```diff
@@ -0,0 +1,104 @@
+  unique-config-entry:
+    status: done
+    comment: |
+      Unique ID is set and duplicates are prevented in homeassistant/components/nrgkick/config_flow.py
```
Please only add comments when there's something exceptional or when exempt or todo

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,777 @@
+            value_path=["info", "general", "rated_current"],
+            precision=2,
+        ),
+        # INFO - Connector
```
Instead, look into entity_descriptions. The goal with entity_descriptions is that you create a generic sensor base class, where you inject an entity_description, and then the sensor will take the details from the entity descriptions. This way you don't have this full list here but outside of the constructor

andijakl commented on homeassistant/components/nrgkick/__init__.py:
```diff
@@ -0,0 +1,56 @@
+
+async def async_reload_entry(hass: HomeAssistant, entry: NRGkickConfigEntry) -> None:
+    """Reload the config entry when it changed."""
+    await hass.config_entries.async_reload(entry.entry_id)
```
This listener is needed because the zeroconf flow can update the host for an existing entry (IP changes). Reloading on entry updates ensures the API client/coordinator are recreated with the updated host (and also applies credential changes from reauth/reconfigure)

andijakl replied:
It would certainly be an edge case. If the device gets a new IP through DHCP, polling would fail. Zeroconf could then automatically update the host on the existing entry. The listener would automatically trigger a reload. Otherwise, the user would have to manually reload to fix the issue. But I'll then remove the listener, as I guess this doesn't happen often!

joostlek replied:
I just checked, we reload automatically
https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L2946-L2994

andijakl commented on homeassistant/components/nrgkick/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    "NRGkickConfigEntry",
+    "NRGkickDataUpdateCoordinator",
+    "NRGkickEntity",
+]
```
True, removing this!

andijakl commented on homeassistant/components/nrgkick/api.py:
```diff
@@ -0,0 +1,234 @@
+
+from .const import DOMAIN
+
+# pylint: enable=import-error
```
Removed (leftover from linting conflicts from community integration)

andijakl committed:
Remove pylint disables, remove re-exporting of classes

andijakl committed:
Remove listener in zeroconf workflow, remove comments from bronze quality scale

andijakl commented on homeassistant/components/nrgkick/entity.py:
```diff
@@ -0,0 +1,57 @@
+        in the UI via translation_key.
+
+        """
+        return self._key
```
This was actually related to an [issue reported by a user in the community integration](https://github.com/andijakl/nrgkick-homeassistant/issues/8), but I'm also a bit sceptical with deviating from the default, so I'll remove that again.

andijakl committed:
Use nrgkick-api library 1.4.0, which includes enums for constant definitions from raw mode of the device

andijakl committed:
Remove fixed English naming for entity IDs and corresponding tests

andijakl committed:
Use dataclass for nrgkick data (info, control, values)

andijakl committed:
Rename ConfigFlow class to NRGkickConfigFlow

andijakl committed:
Remove py.typed as it's not necessary in a core integration

andijakl committed:
Show form error if device does not provide serial, also add new unit test for this case

andijakl commented on homeassistant/components/nrgkick/const.py:
```diff
@@ -0,0 +1,252 @@
+ERROR_GENERAL_OVERTEMPERATURE: Final = 80
+ERROR_HOUSING_OVERTEMPERATURE: Final = 81
+ERROR_ATTACHMENT_OVERTEMPERATURE: Final = 82
+ERROR_DOMESTIC_PLUG_OVERTEMPERATURE: Final = 83
```
Moved to the library (new version 1.4.0, also updated the dependency)

andijakl committed:
Add serial to device info

andijakl committed:
Config flow sets entity.title to device name - now using this instead of setting manually

andijakl committed:
Add enabling JSON api a step in config flow if not already enabled

andijakl committed:
Use decorator for wrapping API calls for more elegant central error transformation

andijakl committed:
Improve handling of user-provided RL in config flow

andijakl committed:
Extend config flow to first try without auth enabled and only prompt the user for auth info if needed

andijakl committed:
Use data-driven entities for sensors

andijakl committed:
Add additional tests for extended config flow, restructure tests

andijakl committed:
New api library version now catches API disabled message from device, raises exception. This is now handled by the config flow of the integration.

andijakl committed:
Fix test that was referencing custom component namespace

andijakl committed:
Fix pylance warnings for tests on accessing config-flow results, ensure HA ui correctly serializes schema parsing info

andijakl committed:
Remove reauth-workflow for first pull request, remove diagnostics, remove remaining code for charging control

andijakl committed:
Update quality scale with current status of the integration

andijakl committed:
Directly return awaited result

andijakl commented:
Thanks for the thorough review - I've implemented all the changes, adapted the automated tests and re-tested with the physical device (also resulted in an updated pypi library). Also rebased the pull request. Ready for review again!

andijakl committed:
Merge branch 'dev' into add-nrgkick-sensor-only

andijakl committed:
Improve tests for disabled JSON API, use raw enums for sensor tests

joostlek commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,413 @@
+    async def async_step_user_auth(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the authentication step only when needed."""
```
Okay so the zeroconf auth step and the zeroconf disabled json api auth step are very similar. In theory you can merge them (like it's fine to merge steps and to have people be guided from the zeroconf flow to the user flow, they won't notice it). Except for one thing and that's the description. If you want a different description you'd have to change the step_id, and implement the `async_step_*` for errors, but you can just `await self.async_step_user_auth` and they'd be redirected back

joostlek commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,413 @@
+                    _LOGGER.exception("Unexpected error")
+                    errors["base"] = "unknown"
+                else:
+                    await self.async_set_unique_id(info["serial"])
```
```suggestion
                    await self.async_set_unique_id(info["serial"], raise_on_progress=False)
```

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        translation_key="connector_serial",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_path=("info", "connector", "serial"),
+    ),
```
this is already stored in the device info, no need to have static data in the state machine

andijakl replied:
Ok sorry! For the connector, we have:
- Phase Count
- Max current
- Type (CEE, Type2, etc.)
- Serial

The max current is also relevant when charging control gets back to the integration, as the HA UI / automations should only allow setting charging values between 6A and the max allowed current (minimum of NRGkick device itself that can be 16/32A and its connector - e.g., domestic 230V plugs shouldn't be used with too high A).

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        value_path=("info", "gps", "accuracy"),
+    ),
```
I think this data is way more usable as a device tracker entity

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        key="charge_pause",
+        translation_key="charge_pause",
+        value_path=("control", "charge_pause"),
+    ),
```
And this sounds like an on/off state. If it's read only -> binary_sensor, otherwise it can be a switch

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfTime.SECONDS,
+        value_path=("values", "general", "vehicle_charging_time"),
+    ),
```
This one is probably different from the previous one as this one can be paused in between and then the timestamp doesn't make sense anymore. However, should we add a suggested unit of measurement as I assume that most people don't read this data in seconds?

andijakl replied:
Yes, pauses could influence the charging time. But a suggested unit of measurement in minutes makes sense (I think hours is a bit too coarse). Thanks for the hint!

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        key="charge_permitted",
+        translation_key="charge_permitted",
+        value_path=("values", "general", "charge_permitted"),
+    ),
```
binary sensor?

andijakl replied:
True, makes sense. Sensor removed to be re-added in a future pull request.

andijakl committed:
Use HA constants for accessing entry data for host, username and password

andijakl committed:
config flow import changes

andijakl commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,413 @@
+                except NRGkickApiClientApiDisabledError:
+                    errors["base"] = "json_api_disabled"
+                except ValueError:
+                    errors["base"] = "no_serial_number"
```
I guess never; but it is a critical value that we need to have, and if the device doesn't return that in its response, I should check if it is present instead of going into a problematic state. I made it as lightweight as possible by not introducing a whole new error class for it but instead using the ValueError; but I think that check should stay here as I have no control over the device. Is that ok?

andijakl commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        value_path=("info", "gps", "accuracy"),
+    ),
```
True, would be good to check. But the [device API](https://nrgkick.com/wp-content/uploads/local_api_docu_simulate-1.html) doesn't have an API to detect if the module is present.
So, I don't know for sure what the device would do if I query it (`/info?gps`) without the SIM/GPS-Option - simply omit the values (most likely), or return empty strings, or return an error? I can't remove that from my device to test, so I think the safest right now is to have it as normal entities, but have them disabled by default. That leaves it with users to know their device and activate these entities if they have the GPS/SIM option and need the data.

andijakl committed:
Remove triggering host retrieval step if not present in later steps and instead assert if set

andijakl committed:
Don't use own unknown state when device api returns unknown, instead default to home assistants own None - unknown state

andijakl committed:
more elegant unknown state handling - still use enum unknown from device api, but set that to home assistant's None

andijakl committed:
shorthand quality scale mapping for done

andijakl committed:
Remove unnecessary comment

andijakl committed:
Remove explicit treatment of missing serial error, treat it as invalid response error instead. Also remove test cases for that scenario

andijakl committed:
Fix doubling of http in help message when user enters full URL for host name in config flow and the host is wrong

andijakl committed:
Simplify entity unique ID during device setup, as serial is required anyway

andijakl committed:
Disable connector serial by default

andijakl committed:
Remove ip address from sensors

andijakl committed:
Add device mac address to device info instead of having it as an entity

andijakl commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        translation_key="network_mac_address",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_path=("info", "network", "mac_address"),
+    ),
```
True, makes sense - done!

andijakl commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        value_path=("info", "versions", "sw_sm"),
+    ),
```
Hm, yes - but what should I then do when the software on the device gets updated? The device info wouldn't regularly be updated automatically. Would it be worth it to always check in the integration if the software version stored in the device info is still up-to-date when the integration queries the device?

joostlek replied:
I would say that depends on how often it updates. If they only update once a year I wouldn't care to be fair, but if they update regularly, you can look into manually updating the value yes. But I'd keep this out for this PR for now, but I have an example in airgradient

andijakl commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        key="phase_count",
+        translation_key="phase_count",
+        value_path=("control", "phase_count"),
+    ),
```
Yes, these 4 values can also be both read & set through the device API to control the charging process.

As such, there should be number entities (and a switch / binary sensor for things like charge pause). However, for the first commit of the new integration, I had to remove these parts to bring it to one platform only.

But having a wallbox HA integration and not seeing some crucial information like if there is an energy limit set wouldn't be very useful I think. Should I transform these 4 entities to number / switch?

joostlek commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        translation_key="charging_rate",
+        state_class=SensorStateClass.MEASUREMENT,
+        value_path=("values", "general", "charging_rate"),
+    ),
```
I think we can use km/h, we shouldn't add the device class though

andijakl committed:
Remove the 4 control-type sensors

andijakl committed:
Add km/h unit of measurement for charging rate sensor

andijakl committed:
Make temperature measurement diagnostic

andijakl committed:
Use enum device class for corresponding sensors

andijakl committed:
Add further info for charge count sensor, and make it diagnostic

andijakl committed:
In entity class, initialize everythin gin constructor. Make assert not just for type checking. No need to retype self.coordinator.data

andijakl committed:
Primary device with empty name will use the entry title by default

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Deactivate raise on progress for async_set_unique_id

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Remove unnecessary constructor for dataclass

andijakl committed:
Move hardware version from sensor entity to device info. Make network ssid disabled by default

andijakl committed:
Remove software version entity as it is already in device info (even though it does not update there, but not so relevant)

andijakl committed:
Made more behind-the-scenes technical energy values disabled by default and some also diagnostic

andijakl committed:
Remove diagnostic type from status and relay state

andijakl commented on homeassistant/components/nrgkick/sensor.py:
```diff
@@ -0,0 +1,734 @@
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_path=("values", "general", "status"),
+        value_fn=lambda value: _map_code_to_translation_key(value, STATUS_MAP),
+    ),
```
True, too important to be only diagnostic. Also changed to enum device class.

andijakl committed:
Report vehicle connected time not in seconds but as timestamp

andijakl committed:
Add suggested unit of measurement in minutes for vehicle charge time

andijakl committed:
Remove GPS sensors as this will be a device tracker platform in a future pull request

andijakl committed:
Remove relay state and grid phases for initial pull request to make them binary sensors in a later release

andijakl committed:
Remove charge_permitted sensor to re-add as binary sensor in a future release

andijakl committed:
Adopt peblar-style API library wrapping instead of own wrapper class

andijakl commented on homeassistant/components/nrgkick/config_flow.py:
```diff
@@ -0,0 +1,413 @@
+            try:
+                host = _normalize_host(user_input[CONF_HOST])
+            except vol.Invalid:
+                errors["base"] = "cannot_connect"
```
Started a quick discussion on discord for that; I added a little HOST_SCHEMA validation, but not much is possible without breaking UI serialization. I think the current approach raising an exception for invalid input is quite nice, as it allows re-using home assistant base classes as much as possible, while still allowing for various user inputs for the hosts (including adding http or not, e.g., adding a trailing slash to the IP)

andijakl committed:
Merge config flow steps for user and password required and json api not enabled

andijakl committed:
Use value_fn instead of getattr for reading sensor values

andijakl committed:
Make config-flow error messages more actionable

andijakl committed:
Make self positional-only to conform with strict mypy typing

andijakl committed:
Merge branch 'dev' into add-nrgkick-sensor-only

andijakl committed:
Add api-retrieved device name for the instantiated nrgkick device

andijakl committed:
Merge branch 'dev' into add-nrgkick-sensor-only

andijakl committed:
Use TextSelector for config flow auth validation

andijakl committed:
Connect to device during zeroconf to catch errors earlier. Update tests to patch these additional scenarios (and refactor to move patching to a common helper function)

andijakl committed:
Remove name from DeviceInfo constructor

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Remove "" around config entry key

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Do not set translations in config flow exceptions, as only errors obj is relevant for the UI

andijakl committed:
Moved validation for non-dict return to nrgkick-api library, updated api version 1.6.0

andijakl committed:
Removed formatting of mac address, as already performed by device registry

andijakl committed:
Remove unnecessary check for None in native_value of sensor.py

andijakl committed:
Remove unknown as already mapped to None

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Remove unknown from further mappings when not required

andijakl committed:
Reword 'error code' to just 'error', and same for 'warning code'

andijakl committed:
Use common strings for the device status

andijakl committed:
Use fixtures for setting up tests, centralize patching into mock_nrgkick_api and also remove patching-function from __init__.py

andijakl committed:
Use more direct and typed assertions

andijakl committed:
Use DOMAIN instead of manually specifying it

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andijakl committed:
Replace hard-coded nrgkick with DOMAIN in tests

andijakl committed:
Remove Test config flow schemas can be serialized for the UI as already tested globally

andijakl committed:
Reordered tests to start with happy path

andijakl committed:
Reuse result variable in tests instead of result2 or result3 to reduce maintenance burden

andijakl committed:
Parametrize tests where it makes sense

andijakl committed:
Re-use mock_config_entry in test_form_already_configured instead of building a new one

andijakl committed:
Remove tests for invalid URL that were not really realistic

andijakl committed:
Move test_validate_input_fallback_name_and_serial_required from helpers to main config flow test file

andijakl committed:
Merge config flow tests into one file

andijakl committed:
Also move host normalization into config flow test file

andijakl committed:
Removed patching of async_get_clientsession, as HA test environment already proivdes valid aiohttp session

andijakl committed:
Add async_setup_integration function to tests to keep commonly used code centralized

andijakl committed:
Remove not needed test_reload_entry test

andijakl committed:
Do not reach into runtime_data, but test through state machine instead

andijakl committed:
Add parametrization to two tests in test_init.py

andijakl committed:
Add snapshot testing for entity values

andijakl committed:
Also introduce snapshot testing for test_sensor.py

andijakl committed:
Improve strict typing for newly added test code

joostlek committed:
Fix

joostlek committed:
Fix

joostlek committed:
Fix

andijakl committed:
Make SIM card device default for tests and snapshots, add description_placeholders for async_step_user

andijakl committed:
Avoid isinstance

andijakl committed:
Remove coordinator decorator pattern and moved the exception handling to the coordinator update method

andijakl committed:
Remove description placeholders from step user
